### PR TITLE
nixos/nixos-install: tell the user what to do if setting a root password failed

### PR DESCRIPTION
**Pull request by Ma27**
###### Motivation for this change

If setting a root password using the `passwd` call in the
`nixos-install` script fails, it should be explained how set it manually
to ensure that nobody gets accidentally locked out of the system.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/installer/tools/nixos-install.sh [hunk 1]
```diff
@@ -138,7 +138,16 @@ fi
...
 # exists (i.e. when mutable user accounts are enabled).
 if [[ -z $noRootPasswd ]] && [ -t 0 ]; then
+    set +e
     nixos-enter --root "$mountPoint" -c '[[ -e /nix/var/nix/profiles/system/sw/bin/passwd ]] && echo "setting root password..." && /nix/var/nix/profiles/system/sw/bin/passwd'
+    exit_code=$?
+    set -e
+
+    if [[ $exit_code != 0 ]]; then
+        echo "Setting a root password failed with the above printed error."
+        echo 'You can set the root password manually by executing `nixos-enter` and then runnig `passwd` in the shell of the new system.'
```
Comment (grahamc) [inline nixos/modules/installer/tools/nixos-install.sh]:
```suggestion
        echo 'You can set the root password manually by executing `nixos-enter` and then running `passwd` in the shell of the new system.'
```

Comment (grahamc) [reply nixos/modules/installer/tools/nixos-install.sh]: Is this always true, or do they need to pass `--root $mountPoint`, and what happens if passwd doesn't exist?

Comment (Ma27) [reply nixos/modules/installer/tools/nixos-install.sh]: You're right, the mount point should be shown as well. And indeed, `nixos-enter` also breaks if the `[[ -e ... ]]` check breaks. I'll fix that :)

**Comment by Ma27** [comment]
@grahamc anything else to add? :)

**Review by matthewbauer — APPROVED** [review]

File: nixos/modules/installer/tools/nixos-install.sh [hunk 2]
```diff
@@ -138,7 +138,18 @@ fi
...
-    nixos-enter --root "$mountPoint" -c '[[ -e /nix/var/nix/profiles/system/sw/bin/passwd ]] && echo "setting root password..." && /nix/var/nix/profiles/system/sw/bin/passwd'
+    if nixos-enter --root "$mountPoint" -c 'test -e /nix/var/nix/profiles/system/sw/bin/passwd'; then
+        set +e
+        nixos-enter --root "$mountPoint" -c 'echo "setting root password..." && /nix/var/nix/profiles/system/sw/bin/passwd'
+        exit_code=$?
+        set -e
+
+        if [[ $exit_code != 0 ]]; then
+            echo "Setting a root password failed with the above printed error."
+            echo "You can set the root password manually by executing \`nixos-enter --root $mountPoint\` and then running \`passwd\` in the shell of the new system."
```
Comment (Infinisil) [inline nixos/modules/installer/tools/nixos-install.sh]: You should use `${mountPoint@Q}` to have it output a working command with proper escaping.

**Comment by Mic92** [comment]
I think we should backport this one.

**Comment by Ma27** [comment]
@Infinisil fixed.
@Mic92 I intended to ask the RMs as soon as this is merged :)

**Review by infinisil — APPROVED** [review]

**Comment by Infinisil** [comment]
Oh and I guess the commit message should be `nixos/nixos-install: ...` by convention

**Comment by Ma27** [comment]
@Infinisil fixed. And thanks for the review!

**Comment by Ma27** [comment]
@lheckemann @samueldr I consider this patch a candidate for backporting. As it touches the NixOS installer, I'm rather hesitant to backport without a review of at least one of you :)

**Comment by Ma27** [comment]
Talked to @lheckemann in person yesterday he told me to go for it with the backport. Now backported as ed21cf98709b370e138009f8315e000848a21870.

**Pull request merged**